aws/amazon-ecs-agent
# Decouple ENI in AttachmentStateChange

### Summary
The problem we've encountered can be summarized with this code: https://github.com/aws/amazon-ecs-agent/blob/master/agent/api/statechange.go#L98-L103
An attachment state change is only ever wrapping the `Attachment *ni.ENIAttachment` type.

In order to save the Resource attachments to our bolt db, we can't keep calling them "ENIs", ([which is our current hack to get around the limitation](https://github.com/aws/amazon-ecs-agent/blob/master/agent/ebs/watcher.go#L317-L319)).

This update puts both the ENI and the Resource attachment types behind an Attachment interface.  Attachment access has moved to use the generic interface methods exposed by the underlying Attachments.  
Note that the bulk of the changed files are to update the import paths.  

### Implementation details
the ecs-agent library `/api` directory now has a new `attachment` type with `eni` and `resource` subdirectories, each containing their attachment type implementations.  All Attachment access is now via the interface methods, except in our boltdb save to disk process for agent restarts which unpacks the generics via type switch and handles the state save/load via separate data handlers for each type.
```
/ecs-agent/api/attachment
    |-> /eni
    |-> /resource
```
Note that the `eniattachment.go` file is still duplicated per https://github.com/aws/amazon-ecs-agent/pull/3801 to support the ongoing common networking library changes.  For now the files are identical except for their package name to allow for the imports to work.  Ultimately once the common library is in place we can remove the duplication.

### Testing
manual test cases:
start an ebs backed task, stop agent/start agent, stop task and see that the stop task cleanup unstage process works.
used a boltdb to json utility to validate the agent.db file contains the attachment payload.

New tests cover the changes: yes

### Description for the changelog
Add generic Attachment interface to allow for multiple attachment types.

**Does this PR include breaking model changes? If so, Have you added transformation functions?**
No this change only includes an additive change -- ResourceAttachments

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
